src/event-db/src/queries/snapshot.rs

```diff
@@ -0,0 +1,122 @@
+use crate::{
+    types::snapshot::{Delegation, Delegator, SnapshotVersion, Voter, VoterInfo},
+    Error, EventDB,
+};
+use async_trait::async_trait;
+
+#[async_trait]
+pub trait SnapshotQueries: Sync + Send + 'static {
+    async fn get_snapshot_versions(&self) -> Result<Vec<SnapshotVersion>, Error>;
+    async fn get_voter(&self, event: String, voting_key: String) -> Result<Voter, Error>;
+    async fn get_delegator(
+        &self,
+        event: String,
+        stake_public_key: String,
+    ) -> Result<Delegator, Error>;
+}
+
+#[async_trait]
+impl SnapshotQueries for EventDB {
+    async fn get_snapshot_versions(&self) -> Result<Vec<SnapshotVersion>, Error> {
+        let conn = self.pool.get().await?;
+        let rows = conn
+            .query(
+                "SELECT event
+                FROM snapshot;",
+                &[],
+            )
+            .await?;
+        let mut snapshot_versions = vec![SnapshotVersion::Latest];
+        for row in rows {
+            let version = SnapshotVersion::Number(row.try_get("event")?);
+            snapshot_versions.push(version);
+        }
+        Ok(snapshot_versions)
+    }
+
+    async fn get_voter(&self, _event: String, voting_key: String) -> Result<Voter, Error> {
+        let conn = self.pool.get().await?;
+        let voter = conn.query_one("SELECT voter.voting_key, voter.voting_group, voter.voting_power, snapshot.as_at, snapshot.last_updated, snapshot.final, SUM(contribution.value)::BIGINT as delegations_power, COUNT(contribution.value) AS delegations_count
+        FROM voter
+        INNER JOIN snapshot ON voter.snapshot_id = snapshot.row_id
+        INNER JOIN contribution ON contribution.snapshot_id = snapshot.row_id
+        WHERE voter.voting_key = $1 AND contribution.voting_key = $1
+        GROUP BY voter.voting_key, voter.voting_group, voter.voting_power, snapshot.as_at, snapshot.last_updated, snapshot.final;", &[&voting_key]).await?;
+
+        let voting_group = voter.try_get("voting_group")?;
+        let voting_power = voter.try_get("voting_power")?;
+
+        let total_voting_power_per_group: i64 = conn
+            .query_one(
+                "SELECT SUM(voter.voting_power)::BIGINT as total_voting_power
+                FROM voter
+                WHERE voter.voting_group = $1;",
+                &[&voting_group],
+            )
+            .await?
+            .try_get("total_voting_power")?;
+
+        Ok(Voter {
+            voter_info: VoterInfo {
+                delegations_power: voter.try_get("delegations_power")?,
+                delegations_count: voter.try_get("delegations_count")?,
+                voting_power_saturation: if total_voting_power_per_group as f64 != 0_f64 {
+                    voting_power as f64 / total_voting_power_per_group as f64
+                } else {
+                    0_f64
+                },
```

Comment:
style nit: can we pull this out into its own calculation just above line 59

Reply:
fixed